File: 2016/day06_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Frequencies = list[dict[str, int]]


def solution(content: list[str]) -> Any:
    frequencies: Frequencies = [{} for _ in range(len(content[0].strip()))]
    for line in content:
        for index, char in enumerate(line.strip()):
            frequencies[index].update({char: frequencies[index].get(char, 0) + 1})
    return "".join(
        sorted(f.items(), key=lambda item: item[1])[0][0] for f in frequencies
    )


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2016/day06-2.ts
type Frequencies = Record<string, number>[];

function solution(content: string) {
  const lines = content.split("\n");
  const frequencies: Frequencies = Array.from({ length: lines[0].length }).map(
    (_) => ({})
  );
  lines.forEach((line) => {
    line.split("").forEach((char, index) => {
      frequencies[index][char] = (frequencies[index][char] ?? 0) + 1;
    });
  });
  return frequencies
    .map(
      (f) =>
        Object.entries(f).toSorted(
          ([_a, aFreq], [_b, bFreq]) => aFreq - bFreq
        )[0][0]
    )
    .join("");
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
